### PR TITLE
core: add hidden layout-shift variant metrics

### PR DESCRIPTION
**Pull request by brendankenny**
Adds the ("normalizing") Layout Shift variants to CLS described in "[Feedback wanted: The road to a better layout shift metric for long-lived pages](https://web.dev/better-layout-shift-metric/)" to the hidden `metrics` audit.

Landing soonish (and doing a release) would allow us to start collecting data over many sites using WPT throttling via HTTP Archive and using simulated throttling through local/cloud testing.

Areas of interest:
- variants vs CLS in lab testing (some speculation that since the window is already limited by load time there may be minimal impact for Lighthouse when switching LayoutShift metrics)
- relative merits of the different variants
- effect of simulated vs applied throttling (do we need some level of Lantern support for the windowed LayoutShift variants?)
- acquire a baseline of data for each variant so if/when one is chosen for a scored metric we'll be able to pick curves and model the effect on scores

Implementation is almost entirely from https://github.com/mmocny/web-vitals/wiki/Snippets-for-LSN-using-PerformanceObserver. Further abstraction/generalization is probably not worth it since we won't be iterating on these in Lighthouse and we expect the multiple-variant implementation to be temporary.

### CODE REVIEW

File: lighthouse-core/computed/layout-shift-variants.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+    return maxScore;
+  }
+
+  /**
+   * Returns all main process LayoutShift events.
+   * @param {LH.Artifacts.TraceOfTab} traceOfTab
+   * @return {Array<LayoutShiftEvent>}
+   */
+  static getLayoutShiftEvents(traceOfTab) {
```
Comment (patrickhulce) [inline lighthouse-core/computed/layout-shift-variants.js:104]: cc @adamraine maybe there's interplay between your planned refactor and this PR? seems like CLS itself could be a "layout shift variant" `maxSession(allEvents, Infinity, Infinity)`

heck our current impl should probably use this :)

File: lighthouse-core/computed/layout-shift-variants.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+    if (layoutShiftEvents.length === 0) return 0;
+
+    let score = 0;
+    let clusterCount = 0;
+    let prevTs = Number.NEGATIVE_INFINITY;
+
+    for (const event of layoutShiftEvents) {
+      if (event.ts - prevTs > 5_000_000) clusterCount++;
+      prevTs = event.ts;
+      score += event.score;
```
Comment (patrickhulce) [inline lighthouse-core/computed/layout-shift-variants.js:39]: are we ignoring framehole here for now?

File: lighthouse-core/test/audits/metrics-test.js [hunk 3]
```diff
@@ -136,4 +136,25 @@ describe('Performance: metrics', () => {
     const result = await MetricsAudit.audit(artifacts, context);
     expect(result.details.items[0].interactive).toEqual(undefined);
   });
+
+  it('evaluates LayoutShift variants correctly', async () => {
+    const artifacts = {
+      traces: {
+        [MetricsAudit.DEFAULT_PASS]: clsAllFramesTrace,
```
Comment (patrickhulce) [inline lighthouse-core/test/audits/metrics-test.js:143]: do we have a real-world test for these where they aren't all identical? if we don't know of any test sites, something that layoutshifts in a set pattern on glitch or something might be just as quick to spin up

File: lighthouse-core/test/computed/layout-shift-variants-test.js [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+        maxSessionGap1s: 3.75,
+        maxSessionGap1sLimit5s: 3.75,
+        maxSliding1s: 3.75,
+        maxSliding300ms: 3.75,
+      });
+    });
+
+    it('includes events with recent input at beginning of trace, but ignores others', async () => {
+      const shiftEvents = [
+        {score: 0.0625, ts: 250_000, had_recent_input: true},
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/layout-shift-variants-test.js:163]: the complex mix of events and scores here made this hard for me to evaluate if it's satisfying the title here. WDYT about `score: 1, had_recent_input: true` for ~5 events, then 1 event 5s later with false then another cluster with true?

File: lighthouse-core/test/computed/layout-shift-variants-test.js [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+        shiftEvents.push({
+          score: 0.125,
+          ts: 1_000_000 + i * 10_000,
+        });
+      }
+      const trace = makeTrace(shiftEvents);
+
+      const variants = await LayoutShiftVariants.request(trace, context);
+      expect(variants).toEqual({
+        avgSessionGap5s: 3.75,
```
Comment (patrickhulce) [inline lighthouse-core/test/computed/layout-shift-variants-test.js:153]:
```suggestion
        avgSessionGap5s: 3.75, // 30 * 0.125
```

**Review by patrickhulce — COMMENTED** [review]
neat!

> some speculation that since the window is already limited by load time there may be minimal impact for Lighthouse when switching LayoutShift metrics

I have a strong suspicion of this as well, at least for the longer windowed variants (and the tests seems to agree 😉 ). Getting this into some FR timespan testing is another story though, big game changer there :D

> effect of simulated vs applied throttling (do we need some level of Lantern support for the windowed LayoutShift variants?)

Do we have reason to suspect this will be any different than for CLS? AFAIK, Lantern CLS is almost always matching or underreporting real CLS and it's essentially impossible to predict the magnitude of the shifts would have occurred when network is stretched out in any principled way.

**Comment by brendankenny** [comment]
> Do we have reason to suspect this will be any different than for CLS?

No, pure--possibly not helpful--speculation :)

File: lighthouse-core/computed/layout-shift-variants.js (hunk 2 above)
Comment (brendankenny) [reply lighthouse-core/computed/layout-shift-variants.js:39]: > are we ignoring framehole here for now?

yes, see the `// Main frame only for now` comment below. Since `cumulativeLayoutShiftAllFrames` is currently in the data gathering stage it seemed like it made the most sense to make the direct comparison to main-frame-only CLS. We could add all frame versions of these variants if it made sense, though?

File: lighthouse-core/test/audits/metrics-test.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/test/audits/metrics-test.js:143]: > do we have a real-world test for these where they aren't all identical? if we don't know of any test sites, something that layoutshifts in a set pattern on glitch or something might be just as quick to spin up

no, closest is `frame-metrics-m89`/`frame-metrics-m90`, but those both only have a single event in the main frame (the other 25 events in `frame-metrics-m89` are in other frames). I didn't really want to add another trace to our menagerie but I guess it's for the best :)

File: lighthouse-core/test/computed/layout-shift-variants-test.js (hunk 4 above)
Comment (brendankenny) [reply lighthouse-core/test/computed/layout-shift-variants-test.js:163]: > the complex mix of events and scores here made this hard for me to evaluate if it's satisfying the title here. WDYT about `score: 1, had_recent_input: true` for ~5 events, then 1 event 5s later with false then another cluster with true?

yeah, the test started as something else and then morphed into this, so that makes sense. SG

File: lighthouse-core/computed/layout-shift-variants.js (hunk 1 above)
Comment (adamraine) [reply lighthouse-core/computed/layout-shift-variants.js:104]: My plan for the refractor was to add the layout shift events for all frames to `TraceOfTab` and then filter/aggregate the events in CLS and CLS-AF. The `had_recent_input` logic would be in `TraceProcessor`.

I could use this `getLayoutShiftEvents` instead if it wasn't specific to main frame events. Maybe pick out the main frame events in `_compute`?  It would also make sense to move this function elsewhere since it wouldn't be specific to this file.

Either way, we should probably be refactoring main frame CLS so we don't repeat the `had_recent_input` logic.

Comment (brendankenny) [reply lighthouse-core/computed/layout-shift-variants.js:104]: > I could use this `getLayoutShiftEvents` instead if it wasn't specific to main frame events. Maybe pick out the main frame events in `_compute`?

yeah, that would work fine

> My plan for the refractor was to add the layout shift events for all frames to `TraceOfTab` and then filter/aggregate the events in CLS and CLS-AF. The `had_recent_input` logic would be in `TraceProcessor`.

@patrickhulce are we just making our job harder for FR, though? I wonder if any correction we do should be pushed back even further...like should the correction be done during gathering, when everything knows whether or not it was a `navigation` and so can confidently correct `had_recent_input`?

OTOH for `timeOrigin` stuff, TraceProcessor is going to have to be given some level of knowledge about what kind of run the trace comes from anyways...

Comment (patrickhulce) [reply lighthouse-core/computed/layout-shift-variants.js:104]: > are we just making our job harder for FR, though? I wonder if any correction we do should be pushed back even further...like should the correction be done during gathering, when everything knows whether or not it was a navigation and so can confidently correct had_recent_input?

When we get there, the workaround will need to be an option for sure. Unfortunately, I think we're just kinda screwed for FR no matter what we do. I called this out in @adamraine's PR too that if we're auto enabling emulation with timespans we'll be stuck in between both of these (we want to ignore had_recent_input for the emulation change just like navigation but the user could have had real input too :/) 

Either way, I'm fine with whatever is most comfortable for these two immediate PRs because FR trace processing is a massive project I don't expect to be significantly affected by this decision here.

> edit: or we just share a getLayoutShiftEvents function between the various CLS impls and don't try to correct the trace at all

This was roughly what I had in mind originally but with `options: {ignoreInitialHadRecentInputSequence: boolean}` or something

File: lighthouse-core/test/audits/metrics-test.js (hunk 3 above)
Comment (brendankenny) [reply lighthouse-core/test/audits/metrics-test.js]: added a trace

**Comment by brendankenny** [comment]
also added tests for boundary conditions, because those are easy to mess up in refactors and not notice

File: lighthouse-core/computed/layout-shift-variants.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/computed/layout-shift-variants.js:39]: I'd like to include the allFrames variants as part of this yeah.

Comment (brendankenny) [reply lighthouse-core/computed/layout-shift-variants.js:39]: > I'd like to include the allFrames variants as part of this yeah.

With the size this PR has grown, would you be ok with a follow-up?

(With #12023 it sounds like we can't do it right until m90, which won't be used by HTTP archive until April (?), so we could maybe wait for @adamraine's refactor as well if that's incoming)

File: lighthouse-core/computed/layout-shift-variants.js (hunk 1 above)
Comment (patrickhulce) [reply lighthouse-core/computed/layout-shift-variants.js:104]: Was this settled? It seems like we're still just duplicating at the moment, and I'm a little unclear of how it will be resolved in a followup.

**Review by patrickhulce — APPROVED** [review]
LGTM

File: lighthouse-core/computed/layout-shift-variants.js (hunk 2 above)
Comment (paulirish) [reply lighthouse-core/computed/layout-shift-variants.js:39]: > > I'd like to include the allFrames variants as part of this yeah.
> 
> With the size this PR has grown, would you be ok with a follow-up?

sure

File: lighthouse-core/computed/layout-shift-variants.js (hunk 1 above)
Comment (brendankenny) [reply lighthouse-core/computed/layout-shift-variants.js:104]: > Was this settled? It seems like we're still just duplicating at the moment, and I'm a little unclear of how it will be resolved in a followup.

this is duplicating but based on https://github.com/GoogleChrome/lighthouse/pull/12034#issuecomment-771889207 (and the conversation before it) it sounded like we still need a third copy and that'll be the time to do the refactor for them to all share some level of implementation. cc @adamraine

Comment (brendankenny) [reply lighthouse-core/computed/layout-shift-variants.js:104]: I can also do the follow up refactor if there's too much else going on right now.

Comment (adamraine) [reply lighthouse-core/computed/layout-shift-variants.js:104]: > it sounded like we still need a third copy and that'll be the time to do the refactor for them to all share some level of implementation

Yeah, I think we should settle the duplication in a follow-up.

Comment (adamraine) [reply lighthouse-core/computed/layout-shift-variants.js:104]: > I can also do the follow up refactor if there's too much else going on right now.

I started working on this locally, but I was waiting for this to land so I could use `getLayoutShiftEvents`.
